File: crates/qow/python/qow/state.py
"""Workflow state management operations.

Thin wrapper around Rust StateManager.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from ._rust import get_engine
from .types import WorkflowState

if TYPE_CHECKING:
    from qo import Queue


async def get_workflow_state(
    _queue: Queue,
    wf_id: str,
) -> tuple[WorkflowState, str]:
    """
    Get workflow state from storage.

    Args:
        _queue: Unused (kept for API compatibility)
        wf_id: Workflow ID

    Returns:
        Tuple of (WorkflowState, etag) for CAS updates

    Raises:
        RuntimeError: If workflow doesn't exist
    """
    engine = await get_engine()
    rust_state, etag = await engine.state.get(wf_id)
    return WorkflowState.from_rust(rust_state), etag


async def update_workflow_state(
    queue: Queue,
    wf_id: str,
    state: WorkflowState,
    etag: str,
    *,
    update_timestamp: bool = True,
) -> str:
    """
    Update workflow state with CAS (compare-and-swap).

    Args:
        queue: The qo queue (used for timestamp)
        wf_id: Workflow ID
        state: New state to write
        etag: Expected etag from previous read
        update_timestamp: Whether to update the updated_at field

    Returns:
        New etag after write

    Raises:
        RuntimeError: If CAS fails (state was modified since read)
    """
    if update_timestamp:
        state.updated_at = await queue.now()

    engine = await get_engine()
    rust_state = state.to_rust()
    return await engine.state.update(wf_id, rust_state, etag, state.updated_at or "")


async def create_workflow_state(
    _queue: Queue,
    wf_id: str,
    state: WorkflowState,
) -> str:
    """
    Create initial workflow state.

    Args:
        _queue: Unused (kept for API compatibility)
        wf_id: Workflow ID
        state: Initial state to write

    Returns:
        Etag of created state

    Note:
        This does not use CAS - it will overwrite existing state.
        Use with caution and idempotency keys.
    """
    engine = await get_engine()
    rust_state = state.to_rust()
    return await engine.state.create(wf_id, rust_state)


async def workflow_exists(
    _queue: Queue,
    wf_id: str,
) -> bool:
    """Check if a workflow exists."""
    engine = await get_engine()
    return await engine.state.exists(wf_id)


async def delete_workflow_state(
    _queue: Queue,
    wf_id: str,
) -> None:
    """
    Delete workflow state (for cleanup/testing).

    Args:
        _queue: Unused (kept for API compatibility)
        wf_id: Workflow ID
    """
    engine = await get_engine()
    await engine.state.delete(wf_id)


async def save_step_result(
    _queue: Queue,
    wf_id: str,
    step_name: str,
    result: bytes,
) -> str | None:
    """
    Save step result to storage (atomic, idempotent - won't overwrite existing).

    Uses S3 if-none-match for atomic writes.

    Args:
        _queue: Unused (kept for API compatibility)
        wf_id: Workflow ID
        step_name: Name of the completed step
        result: Serialized step result (JSON bytes)

    Returns:
        Etag of saved result, or None if result already exists
    """
    engine = await get_engine()
    return await engine.state.save_step_result(wf_id, step_name, result)


async def get_step_result(
    _queue: Queue,
    wf_id: str,
    step_name: str,
) -> tuple[bytes, str] | None:
    """
    Get step result from storage.

    Args:
        _queue: Unused (kept for API compatibility)
        wf_id: Workflow ID
        step_name: Name of the step

    Returns:
        Tuple of (result bytes, etag) or None if not found
    """
    engine = await get_engine()
    return await engine.state.get_step_result(wf_id, step_name)


async def list_workflows(
    _queue: Queue,
    prefix: str = "",
    limit: int = 1000,
) -> list[str]:
    """
    List workflow IDs.

    Args:
        _queue: Unused (kept for API compatibility)
        prefix: Optional prefix to filter workflows
        limit: Maximum number of workflows to return

    Returns:
        List of workflow IDs
    """
    engine = await get_engine()
    return await engine.state.list(prefix, limit)

File: crates/buquet-workflow/src/state.rs
//! Workflow state management operations.
//!
//! This module provides state management for workflows stored in S3.
//! State is stored at `workflow/{wf_id}/state.json` and step results
//! are stored at `workflow/{wf_id}/steps/{step_name}.json`.

use crate::error::{Result, WorkflowError};
use crate::types::WorkflowState;
use buquet::storage::{PutCondition, S3Client};
use std::sync::Arc;

/// Get the S3 key for workflow state.
fn state_key(wf_id: &str) -> String {
    format!("workflow/{wf_id}/state.json")
}

/// Get the S3 key for step result.
fn step_result_key(wf_id: &str, step_name: &str) -> String {
    format!("workflow/{wf_id}/steps/{step_name}.json")
}

/// Manages workflow state in S3.
///
/// Provides operations for creating, reading, updating, and deleting
/// workflow state with support for optimistic concurrency control via `ETags`.
#[derive(Debug, Clone)]
pub struct StateManager {
    client: Arc<S3Client>,
}

impl StateManager {
    /// Create a new `StateManager` with the given S3 client.
    #[must_use]
    pub const fn new(client: Arc<S3Client>) -> Self {
        Self { client }
    }

    /// Get workflow state from storage.
    ///
    /// Returns the workflow state and its `ETag` for use in CAS updates.
    ///
    /// # Errors
    ///
    /// Returns `WorkflowError::NotFound` if the workflow doesn't exist.
    pub async fn get(&self, wf_id: &str) -> Result<(WorkflowState, String)> {
        let key = state_key(wf_id);
        let (data, etag) = self.client.get_object(&key).await.map_err(|e| {
            if matches!(e, buquet::storage::StorageError::NotFound { .. }) {
                WorkflowError::NotFound(wf_id.to_string())
            } else {
                WorkflowError::Storage(e)
            }
        })?;

        let state = WorkflowState::from_json(&data)?;
        Ok((state, etag))
    }

    /// Update workflow state with CAS (compare-and-swap).
    ///
    /// The update will only succeed if the current `ETag` matches the provided
    /// `etag`. The `now` timestamp is used to update the `updated_at` field.
    ///
    /// # Arguments
    ///
    /// * `wf_id` - Workflow ID
    /// * `state` - New state to write
    /// * `etag` - Expected `ETag` from previous read
    /// * `now` - Current timestamp to set as `updated_at`
    ///
    /// # Returns
    ///
    /// New `ETag` after successful write.
    ///
    /// # Errors
    ///
    /// Returns `WorkflowError::Conflict` if the `ETag` doesn't match (state was
    /// modified by another process).
    pub async fn update(
        &self,
        wf_id: &str,
        state: &WorkflowState,
        etag: &str,
        now: &str,
    ) -> Result<String> {
        // Clone and update timestamp
        let mut state = state.clone();
        state.updated_at = now.to_string();

        let key = state_key(wf_id);
        let data = state.to_json()?;

        let new_etag = self
            .client
            .put_object(&key, data, PutCondition::IfMatch(etag.to_string()))
            .await
            .map_err(|e| {
                if matches!(e, buquet::storage::StorageError::PreconditionFailed { .. }) {
                    WorkflowError::Conflict
                } else {
                    WorkflowError::Storage(e)
                }
            })?;

        Ok(new_etag)
    }

    /// Create initial workflow state.
    ///
    /// This does not use CAS - it will overwrite existing state.
    /// Use with caution and ensure idempotency at a higher level.
    ///
    /// # Returns
    ///
    /// `ETag` of the created state.
    pub async fn create(&self, wf_id: &str, state: &WorkflowState) -> Result<String> {
        let key = state_key(wf_id);
        let data = state.to_json()?;

        let etag = self
            .client
            .put_object(&key, data, PutCondition::None)
            .await?;

        Ok(etag)
    }

    /// Create workflow state only if it doesn't already exist.
    ///
    /// Uses `IfNoneMatch` condition for idempotent creation.
    ///
    /// # Returns
    ///
    /// `Ok(Some(etag))` if created, `Ok(None)` if already exists.
    pub async fn create_if_not_exists(
        &self,
        wf_id: &str,
        state: &WorkflowState,
    ) -> Result<Option<String>> {
        let key = state_key(wf_id);
        let data = state.to_json()?;

        match self
            .client
            .put_object(&key, data, PutCondition::IfNoneMatch)
            .await
        {
            Ok(etag) => Ok(Some(etag)),
            Err(buquet::storage::StorageError::PreconditionFailed { .. }) => Ok(None),
            Err(e) => Err(WorkflowError::Storage(e)),
        }
    }

    /// Check if a workflow exists.
    pub async fn exists(&self, wf_id: &str) -> Result<bool> {
        let key = state_key(wf_id);
        let exists = self.client.head_object(&key).await?;
        Ok(exists)
    }

    /// Delete workflow state.
    ///
    /// This is primarily for cleanup and testing. In production, consider
    /// archiving workflows instead of deleting them.
    pub async fn delete(&self, wf_id: &str) -> Result<()> {
        let key = state_key(wf_id);
        self.client.delete_object(&key).await?;
        Ok(())
    }

    /// List workflow IDs with an optional prefix filter.
    ///
    /// # Arguments
    ///
    /// * `prefix` - Optional prefix to filter workflows (appended to "workflow/")
    /// * `limit` - Maximum number of workflows to return
    ///
    /// # Returns
    ///
    /// List of workflow IDs (not full keys).
    pub async fn list(&self, prefix: &str, limit: i32) -> Result<Vec<String>> {
        let search_prefix = format!("workflow/{prefix}");
        let keys = self
            .client
            .list_objects_paginated(&search_prefix, limit)
            .await?;

        // Extract workflow IDs from state.json keys
        // Key format: workflow/{wf_id}/state.json or workflow/{wf_id}/steps/...
        let mut workflow_ids = std::collections::HashSet::new();
        for key in keys {
            let parts: Vec<&str> = key.split('/').collect();
            if parts.len() >= 2 {
                workflow_ids.insert(parts[1].to_string());
            }
        }

        let mut ids: Vec<String> = workflow_ids.into_iter().collect();
        ids.sort();
        Ok(ids)
    }

    /// Save step result to storage.
    ///
    /// Step results are immutable after completion and stored separately
    /// from workflow state for easier inspection and potential payload
    /// reference support.
    ///
    /// # Arguments
    ///
    /// * `wf_id` - Workflow ID
    /// * `step` - Name of the completed step
    /// * `result` - Serialized step result (JSON bytes)
    ///
    /// # Returns
    ///
    /// `Ok(Some(etag))` if created, `Ok(None)` if result already exists.
    pub async fn save_step_result(
        &self,
        wf_id: &str,
        step: &str,
        result: &[u8],
    ) -> Result<Option<String>> {
        let key = step_result_key(wf_id, step);
        match self
            .client
            .put_object(&key, result.to_vec(), PutCondition::IfNoneMatch)
            .await
        {
            Ok(etag) => Ok(Some(etag)),
            Err(buquet::storage::StorageError::PreconditionFailed { .. }) => Ok(None),
            Err(e) => Err(WorkflowError::Storage(e)),
        }
    }

    /// Get step result from storage.
    ///
    /// # Returns
    ///
    /// Tuple of (result bytes, `ETag`) or `None` if not found.
    pub async fn get_step_result(
        &self,
        wf_id: &str,
        step: &str,
    ) -> Result<Option<(Vec<u8>, String)>> {
        let key = step_result_key(wf_id, step);
        match self.client.get_object(&key).await {
            Ok((data, etag)) => Ok(Some((data, etag))),
            Err(buquet::storage::StorageError::NotFound { .. }) => Ok(None),
            Err(e) => Err(WorkflowError::Storage(e)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_state_key() {
        assert_eq!(state_key("wf-123"), "workflow/wf-123/state.json");
    }

    #[test]
    fn test_step_result_key() {
        assert_eq!(
            step_result_key("wf-123", "process_order"),
            "workflow/wf-123/steps/process_order.json"
        );
    }
}
